[AIRFLOW-5073] Treat NULL as keep poking in SQL sensor

uncletoxa opened the pull request:
Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following:
  - https://issues.apache.org/jira/browse/AIRFLOW-5073


### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
- switches NULL response to 'failed' result instead of 'success'
- adds `allow_true` parameter to preserve previous behaviour

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- tests are extremely trivial, but I can add them if needed
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

uncletoxa committed:
[AIRFLOW-5073] Treat NULL as fail in SQL sensor

Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

uncletoxa committed:
[AIRFLOW-5073] Treat NULL as fail in SQL sensor

Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

pgagnon reviewed (commented):
I think you should add a test even though it's trivial, if only so people do not break your feature in the future.

uncletoxa committed:
[AIRFLOW-5073] Treat NULL as fail in SQL sensor

Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

uncletoxa committed:
Merge branch 'master' into master

ashb reviewed (changes requested):
Yup, please keep the default behaviour unchanged.

Oh you have. We don't need the entry in UPDATING then

dstandish commented on UPDATING.md:
```diff
@@ -23,6 +23,12 @@ assists users migrating to a new version.
+### Changes to SqlSensor
+
+`'None'` is included in success criteria along with `0`, `'0'` and `''`. To preserve 
+previous behaviour `allow_true` boolean parameter is introduced. It will exclude `'None'`
```
>`'None'` is included in success criteria along with `0`, `'0'` and `''`. To preserve 

What you're actually doing is including `None` in "keep poking" criteria.

Previous behavior line is not worded clearly.  It excludes `None` from what?  Given what value?

Should be `allow_null` not `allow_true`

dstandish commented on airflow/sensors/sql_sensor.py:
```diff
@@ -28,9 +28,10 @@
-    success or failure criteria are met, or if the first cell is in (0, '0', ''). Optional success
-    and failure callables are called with the first cell returned as the argument. If success
-    callable is defined the sensor will keep retrying until the criteria is met.
+    success or failure criteria are met, or if the first cell is in (0, '0', '', 'None').
```
> It will keep trying until success or failure criteria are met, or if the first cell is in (0, '0', '', 'None').

This could be made more clear.  Shouldn't it be "first cell is _not_ in ..."?

dstandish commented on airflow/sensors/sql_sensor.py:
```diff
@@ -28,9 +28,10 @@
-    and failure callables are called with the first cell returned as the argument. If success
-    callable is defined the sensor will keep retrying until the criteria is met.
+    success or failure criteria are met, or if the first cell is in (0, '0', '', 'None').
+    An allow_null parameter exclude 'None' results from failure criteria.
```
> An allow_null parameter exclude 'None' results from failure criteria.

'None' was never part of failure criteria; `None` was implicitly part of the default _success_ criteria, since it was not on the "keep poking" list, and success is defined by negation.

dstandish reviewed (changes requested):
> don't you think changing the default behavior would be a good idea?

So, disclosure, I don't use sql sensor. But here are my thoughts.
I agree the current design is quite awkward: 0 and '' mean "not done", yet NULL means "done".

Having looked into this more, if anything I'd actually vote do _not_ add `allow_null` parameter; just add `None` to the "Not done" list, i.e. as though this were a bugfix. The sensor is already confusing enough and if someone really needs None to mean "done" they can add a success callable e.g. `lambda x: str(x) == 'None'`. But I'm not well versed in norms re changing behavior in a project like this.

In any case, I think it is misleading / confusing to say "treat null as fail". In the context of sensors, I think fail means raise exception. E.g. if the failure callback evaluates to True then an exception is raised. What we're talking about here though is the "keep poking" list, i.e. whether a `None` results in a True or False, i.e. success or keep poking. Maybe this is nit picky but I think something like "Do not interpret NULL as success" would be more accurate.

uncletoxa committed:
[AIRFLOW-5073] Treat NULL as keep poking in SQL sensor

Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

ashb commented on airflow/sensors/sql_sensor.py:
```diff
@@ -98,4 +102,6 @@ def poke(self, context):
-        return str(first_cell) not in ('0', '')
+        if self.allow_null:
+            return str(first_cell) not in ('0', '')
+        return str(first_cell) not in ('0', '', 'None')
```
```suggestion
        return first_cell is not None and str(first_cell) not in ('0', '')
``

ashb replied:
(And although not  a new bug, this won't work if the type is floating point, that would give us 0.0 as a value)
